### ISSUE TITLE
Rule proposal: `prefer-string-raw`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#2048]
### Description

To avoid escape `\`.

### Fail

```js
console.log('"\\" found.');
```

```js
const directory = "C:\\windows\\style\\path";
```

```js
new RegExp('foo\\.bar')
```

### Pass

```js
console.log(String.raw`"\" found.`);
```

```js
const directory = String.raw`C:\windows\style\path`;
```

```js
new RegExp(String.raw`foo\.bar`)
```

```js
const foo = '\\ `';
// Use `String.raw` will need escape `
const foo = '\\ ${foo}';
// Use `String.raw` will need escape $ or `{`
```

```js
// Places can't use a tagged template

"A directive\\"

import foo from "./foo\\bar.js";

const foo = {
	'\\': 'a non-computed key'
}
```

---

This rule can also check when building a tagged function like [this](https://github.com/fregante/code-tag/commit/b20054ce644edb85f17a93ec364371bc3943455f), but it's hard to detect such pattern.

**Comment by silverwind** [comment]
Seems useful sometimes. I think the rule should not trigger when the string in question contains other escape sequences besides `\\` because using `String.raw` on that would break those other sequences, for example:

```
console.log(`"\\" found. \nok`);
```

**Comment by fisker** [comment]
@sindresorhus Thoughts?

**Comment by sindresorhus** [comment]
Accepted

**Comment by DavidAnson** [comment]
@fisker @sindresorhus, is this rule compelling enough to have in the “recommended” set? I normally try to abide by plugin defaults, but I’m struggling here. While C#/Python syntax for raw is short, JavaScript syntax is long and ugly. This is personal preference, but the examples in the documentation feel to me like they are worse after the recommended change. Also, if the reader is not familiar with template literals, the JS syntax looks wrong/invalid. According to mdn, raw is the “only built-in template literal tag” which makes me think it won’t be familiar to a lot of people. To be clear, I don’t object to the rule itself, I’m just questioning whether it should be recommended?

**Comment by fisker** [comment]
> the examples in the documentation feel to me like they are worse after the recommended change.

Can you share an example?

**Comment by fisker** [comment]
> the examples in the documentation feel to me like they are worse after the recommended change

Why is that? Escaping `\` is ugly to me.

**Comment by DavidAnson** [comment]
The `console.log` and `RegExp` examples are almost twice as long after adding `String.raw` - and the first/most important parameter to both functions is partially obscured because it's prefixed by `String.raw`.

**Comment by silverwind** [comment]
I tired to enable this rule, but ended up disabling it because the only cases where I use backslashes is in RegExp strings like `"[^\\/]"`, and I found the readability reduced because `String.raw` is not well known and the intend of avoiding to escape a single character is not obvious.

**Comment by ElPrudi** [comment]
Should not be in recommended. This kills performance and VSCode code syntax doesn't know how to handle it:

```js
console.time('before')
for (let i = 0; i < 100000; i++) 'C:\\windows\\style\\path'
console.timeEnd('before') // 1.935 ms
console.time('after')
for (let i = 0; i < 100000; i++) String.raw`C:\windows\style\path`
console.timeEnd('after') // 47.016 ms
// Around: 25x slower
```
